cdap-docs/reference-manual/source/http-restful-api/metadata.rst

```diff
@@ -1302,3 +1333,55 @@ with the metadata returned as a JSON string in the return body::
      - The properties requested were returned as a JSON string in the body of the response
    * - ``404 NOT FOUND``
      - The entity, program, or run for which properties are being requested was not found
+
+**Note**: The metadata aggregation of program, application, stream and dataset associated with a program is
+deprecated and program runs are treated like any other entity to which metadata can be associated directly.
+To retrieve the direct metadata of program run please refer to
+:ref:`Metadata Properties <http-restful-api-metadata-properties>` and
+:ref:`Metadata Tags <http-restful-api-metadata-tags>`. An additional query parameter ``aggregateRun`` must be set to
+``false`` while operating with metadata directly associated with program runs.
+
+Metadata for Custom Entities
+============================
+
+Metadata can also be associated with custom entities. In CDAP entities are separated into two kinds:
+
+- CDAP Entities: These entities are entities which are defined by CDAP and have special meaning in CDAP. ``Namespace``,``Stream``, ``Application``, ``Dataset`` etc. are example of CDAP Entities.
+
+- Custom Entities: These entities are user defined entities to represent a resource which exists in CDAP. Custom Entities are represented as a hierarchical key-value pair and can optionally have a explicitly defined type.
+
+If a type is not specified then the last key in the hierarchy is considered as the type.
+
+In REST APIs the custom entities can be represented as the hierarchical key-value pair and an optional type as query
+parameter if the last key in the hierarchy is not the type.
+
+For example to add tags to a file in a PartitionFileSet::
+
+  POST /v3/namespaces/<namespace-id>/datasets/<dataset-id>/file/<file-name>/metadata/tags
+
+or, to a jar present in a namespace::
+
+  POST /v3/namespaces/<namespace-id>/jar/<jar-id>/versions/<jar-version>/metadata/tags[?type=<type>
+
```

Comment:
In the example above, the custom entity consists of two key-value pairs. The first has key `jar` and value `<jar-id>`. The second has key `versions` and value `<jar-version>`.

(what's the type in this custom entity?)

Reply:
Done.